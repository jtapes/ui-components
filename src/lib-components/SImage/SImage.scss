@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-image {
  position: relative;
  width: 100%;
  height: 100%;

  &__inner {
    width: 100%;
    height: 100%;
  }

  &__tags {
    position: absolute;
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    padding: $size--16;
    max-width: 100%;

    .s-tag {
      padding: $size--4 $size--8;
    }

    &--top-left {
      left: 0;
      top: 0;
    }

    &--top-right {
      right: 0;
      top: 0;
    }

    &--bottom-left {
      left: 0;
      bottom: 0;
    }

    &--bottom-right {
      right: 0;
      bottom: 0;
    }
  }
}
