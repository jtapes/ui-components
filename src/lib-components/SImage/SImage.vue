<template>
  <div class="s-image" :style="wrapperSize">
    <div
      v-if="$slots.tags"
      class="s-image__tags"
      :class="[
        tagContainerClass,
        `s-image__tags--${tagsPlacement}`
      ]"
    >
      <!-- @slot Слот для тегов -->
      <slot name="tags" />
    </div>
    <!-- @slot Слот для контена при ошибке загрузки изображения -->
    <slot v-if="$slots.error && isError" name="error" />
    <img
      v-if="!isError || (!$slots.error && isError)"
      ref="image"
      class="s-image__inner"
      :src="src"
      :loading="lazy ? 'lazy' : 'eager'"
      :alt="alt"
      :style="imageStyle"
    />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

enum TagPlacements {
  TOP_LEFT = 'top-left',
  TOP_RIGHT = 'top-right',
  BOTTOM_LEFT = 'bottom-left',
  BOTTOM_RIGHT = 'bottom-right'
}

/** Компонент для отображения фотографии */
@Component
export default class SImage extends Vue {
  isError = false

  /** Путь к изображению */
  @Prop({ type: String, default: '/images/no-photo.svg' })
  src!: string

  /** Альтернативный текст изображения */
  @Prop(String)
  alt?: string

  /** Альтернативный текст изображения */
  @Prop(String)
  fit?: string

  /** Ленивая загрузка (default = false) */
  @Prop({ type: Boolean, default: false })
  lazy!: boolean

  /** Ширина изображения */
  @Prop([String, Number])
  width?: string | number

  /** Высота изображения */
  @Prop([String, Number])
  height?: string | number

  /** Класс для контейнера тегов */
  @Prop(String)
  tagContainerClass?: string

  /** Расположение тегов ('top-left' (default) or 'top-right' or 'bottom-left' or 'bottom-right')  */
  @Prop({
    type: String,
    default: 'top-left',
    validator(placement: string): boolean {
      return Object.values(TagPlacements).includes(placement as TagPlacements)
    }
  })
  tagsPlacement!: string

  get wrapperSize(): object {
    return {
      width: Number.isInteger(this.width) ? `${this.width}px` : this.width,
      height: Number.isInteger(this.height) ? `${this.height}px` : this.height,
    }
  }

  get imageStyle(): object {
    return { 'object-fit': this.fit }
  }

  handleError(): void {
    const image = this.$refs.image as HTMLImageElement
    image.src = '/images/no-photo.svg'
    this.isError = true
  }

  mounted(): void {
    const image = this.$refs.image as EventTarget
    image.addEventListener('error', this.handleError, {
      once: true,
    })
  }

  beforeDestroy(): void {
    const image = this.$refs.image as EventTarget
    image.removeEventListener('error', this.handleError)
  }
}
</script>
