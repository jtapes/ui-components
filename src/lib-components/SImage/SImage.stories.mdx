import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest'
import {
  Meta,
  Story,
  Source,
  Canvas,
  ArgsTable,
  Description,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SImage from '@/lib-components/SImage/SImage.vue'
import STag from '@/lib-components/STag/STag.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Image"
  component={SImage}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    src: { control: 'text' },
    alt: { control: 'text' },
    width: { control: 'text' },
    height: { control: 'text' },
    fit: {
      options: [
        null,
        'contain',
        'cover',
        'none',
        'scale-down',
        'fill',
      ],
      control: { type: 'select' },
      defaultValue: null,
    },
    tagContainerClass: { control: 'text' },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-%2F-Market-%2F-Catalog-%E2%80%A2-Product-Page?node-id=9%3A3',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SImage },
  template: `
    <s-image :src="src" :alt="alt" :width="width" :height="height" :fit="fit" :tag-container-class="tagContainerClass" />
  `,
})

export const TemplateWithTag = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SImage, STag },
  template: `
    <s-image :src="src" :alt="alt" width="200px" height="250px" fit="scale-down" :tag-container-class="tagContainerClass">
      <template #tags>
        <s-tag label="Выгодно" bg-color="mosque" text-color="white"/>
        <s-tag label="-12%" bg-color="danger" text-color="white"/>
      </template>
    </s-image>
  `
})

# Изображение

<Description />

<Canvas withToolbar withSource="none">
  <Story name="SImage">
    {Template.bind({})}
  </Story>
  <Story name="SImage(with tag)">
    {TemplateWithTag.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`
    <s-image src="example.png" alt="Изображение" width="50%" height="50%" fit="scale-down" />
  `}
/>

<ArgsTable />

<div style={{padding: '10px 0', marginBottom: '15px'}}>
  <span style={{color: '#111111', fontWeight: 700, fontSize: '20px'}}>
    В атрибуты <span style={{color: '#606060', fontSize: '20px'}}>width</span> и <span style={{color: '#606060', fontSize: '20px'}}>height</span> можно передавать как строковые значения
  </span>
  <span style={{color: '#606060', fontSize: '18px', marginLeft: '5px'}}>(100px / 3rem / 50% и т.п.)</span>
  <span style={{color: '#111111', fontWeight: 700, fontSize: '20px'}}>, так и числовые.</span>
  <div style={{padding: '20px'}}>
    <span>Числовые значения будут преобразованы в единицу измерения "пиксели" (px).</span>
  </div>
</div>
