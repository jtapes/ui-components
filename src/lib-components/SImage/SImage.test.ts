import { mount } from '@vue/test-utils'
import SImage from '@/lib-components/SImage/SImage.vue'

describe('SImage', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SImage)
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Error load image', () => {
    const wrapper = mount(SImage, {
      propsData: {
        width: 200,
        height: 200,
        src: '/images/test-image.png',
      },
    })
    const image = wrapper.find('img')
    image.trigger('error')
    expect(wrapper.find('img').attributes('src')).toBeTruthy()
    wrapper.destroy()
  })
})
