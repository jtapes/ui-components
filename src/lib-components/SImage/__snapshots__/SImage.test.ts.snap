// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SImage Is Vue snapshot 1`] = `
<div
  class="s-image"
>
  <!---->
   
  <!---->
   
  <img
    class="s-image__inner"
    loading="eager"
    src="/images/no-photo.svg"
  />
</div>
`;
